plugins/core/src/terminal-service.ts

```
@@ -0,0 +1,73 @@
+        const cp: IPty = spawn(process.env.SHELL, [], {});
+        const queue = createAsyncQueue<Buffer>();
+        cp.onExit(() => queue.end());
+
+        let bufferedLength = 0;
+        const MAX_BUFFERED_LENGTH = 64000;
+        cp.onData(async data => {
+            const buffer = Buffer.from(data);
+            bufferedLength += buffer.length;
+            const promise = queue.enqueue(buffer).then(() => bufferedLength -= buffer.length);
+            if (bufferedLength >= MAX_BUFFERED_LENGTH) {
+                cp.pause();
+                await promise;
+                if (bufferedLength < MAX_BUFFERED_LENGTH)
+                    cp.resume();
+            }
+        });
+
+        async function* generator() {
+            try {
+                while (true) {
+                    const buffers = queue.clear();
+                    if (buffers.length) {
+                        yield Buffer.concat(buffers);
+                        continue;
+                    }
+
+                    yield await queue.dequeue();
+                }
+            }
+            finally {
+                cp.kill();
+            }
+        }
+
+        setTimeout(async () => {
+            try {
+                for await (const message of input) {
+                    if (!message) {
+                        cp.kill();
+                        return;
+                    }
+                    if (Buffer.isBuffer(message)) {
+                        cp.write(message.toString());
+                        continue;
+                    }
+                    try {
+                        const parsed = JSON.parse(message.toString());
+                        if (parsed.dim) {
+                            cp.resize(parsed.dim.cols, parsed.dim.rows);
+                        }
+                    } catch {
+                        cp.write(message.toString());
+                    }
+                }
+            }
+            finally {
```

Should catch and eat all errors here.